### PR TITLE
feat(moe): Support placing MTP layers into standalone stages

### PR DESCRIPTION
**Pull request by BestJuly**
# What does this PR do ?
Support placing MTP to a standalone stage, e.g., MTP in the 2nd last VPP stage for better VPP balance.
This is the PR for main branch, and the corresponding [PR1916](https://github.com/NVIDIA/Megatron-LM/pull/1916) for dev branch.

## Highlights
This document outlines the design of a new MTP (Multi-Token Prediction) standalone feature. This feature allows MTP layers to be placed into a standalone VPP (Virtual Pipeline Parallelism) stage, rather than being confined to the last VPP stage.

### Key Benefits:
- Enhanced Flexibility in MTP VPP Placement
  - MTP layers can now be positioned in either the last or second-to-last VPP stage, offering greater control over pipeline configuration.
- Improved Load Balancing
  - The MTP standalone feature isolates transformer block-related computations, which have a similar computational cost to a normal transformer block. This split, with MTP loss calculation remaining in the second-to-last VPP stage, contributes to a more balanced workload.
- Minimal Impact on Existing Code Paths
  - The design aims for near-zero impact on current code. A minor change involves the final layernorm layer, which will now be placed at the end of the decoder stage. In scenarios where the last stage contains only MTP layers and loss, this layernorm will reside in the second-to-last layer alongside a decoder layer. While this differs from previous implementations (where the final layernorm was always in the last stage), bitwise correctness has been validated under various conditions, including forceful layernorm placement and disabled gradient clipping.

### Considerations:
- Changes in P2P Communication Shapes
  - The need to pass original and MTP hidden states to subsequent VPP stages will alter p2p communication shapes. To mitigate the impact on core communication, the shape is communicated upfront when MTP standalone settings are detected in the pipeline layout.

## Design idea
We split MTP computation into two parts
- Output head computation.
- Others.   
image

The advantage is that we put the MTP loss computation as main, so for the untie embedding part, the MTP loss can use exactly the same code path of the lm loss as they are in the same vpp stage, which is simpler than putting MTP loss calculation in the same vpp stage as MR#2996 so that we can avoid bugs related to this. With the new design, the DSv3 TFLOPs will also be compatible with the #2996 one which included in the 0.13 EA branch.We divide MTP computation into two parts: output head computation and others. This approach offers the advantage of treating MTP loss computation as primary. Consequently, for the untied embedding, the MTP loss can utilize the exact same code path as the LM loss, as they are within the same VPP stage. This simplifies the process compared to placing MTP loss calculation in the same VPP stage as MR#2996, thereby mitigating potential bugs. Furthermore, this new design ensures DSv3 TFLOPs compatibility with the #2996 version included in the 0.13 EA branch.

### Perf
- The throughput in DSv3 is on par or even higher than 0.13 EA (where we use previous design for this standalone feature).

### Convergence
**On full DSv3**   
(Real training mode) Almost overlapped with the main branch settings for the metrics which is related to convergence.

**Deterministic correctness check**
We validate the correctness in mainly two parts (when using different pp layout, can be bitwise aligned when manually disabling grad clipping)
- Deterministic alignment between main branch and this MR.
  - The same pipeline layout settings;
  - Different pipeline layout settings.
- Deterministic alignment within this MR to validate the compatibility of 1F1B overlap.

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Review by yanring — APPROVED** [review]
Already reviewed on the dev PR, approve now.

File: megatron/core/datasets/blended_megatron_dataset_builder.py [hunk 1]
```diff
@@ -47,11 +47,13 @@ def __init__(
         sizes: List[int],
         is_built_on_rank: Callable,
         config: BlendedMegatronDatasetConfig,
+        vp_stage: Optional[int] = None,
     ):
         self.cls = cls
         self.sizes = sizes
         self.is_built_on_rank = is_built_on_rank
         self.config = config
+        self.vp_stage = vp_stage
```
Comment (asolergi-nv) [inline megatron/core/datasets/blended_megatron_dataset_builder.py]: Why is this needed here? We are not even passing the `vp_stage` when calling the  `BlendedMegatronDatasetBuilder`

https://github.com/BestJuly/Megatron-LM/blob/c1022a7696393759729dbb8da3ff7144f27df634/pretrain_gpt.py#L230-L233

**Comment by kvareddy** [comment]
@deepakn94 can you please take a look at this MR?

File: megatron/core/datasets/blended_megatron_dataset_builder.py (hunk 1 above)
Comment (jkamalu) [reply megatron/core/datasets/blended_megatron_dataset_builder.py]: I am also confused. Is it because we want to make sure we don't build the datasets on more than one vp stage per rank?

Comment (BestJuly) [reply megatron/core/datasets/blended_megatron_dataset_builder.py]: This should be wrongly added when resolving conflicts, as I find vp_stage has been included in `partial(is_dataset_built_on_rank, vp_stage=vp_stage)`. Will correct this.

**Review by ko3n1g — APPROVED** [review]

**Comment by Phlip79** [comment]
@BestJuly can you please fix conflicts?

File: megatron/core/transformer/pipeline_parallel_layer_layout.py [hunk 2]
```diff
@@ -127,15 +127,28 @@ def validate_layer_layout(self, num_layers: int, mtp_num_layers: int):
...
                     self.layout[pp_rank][-1].count(LayerType.mtp) == mtp_num_layers
-                ), "All of the MTP layers must be in the same stage"
-                assert (
-                    pp_rank == self.pipeline_model_parallel_size - 1
-                    and LayerType.loss in self.layout[pp_rank][-1]
-                ), "MTP layers must be in the last stage together with Loss stage."
+                ), "All of the MTP layers must be in the same one virtual pipeline stage"
+        for vpp_rank in range(self.virtual_pipeline_model_parallel_size - 1):
+            assert LayerType.mtp not in self.layout[0][vpp_rank], (
+                f"Corrently we restrict that the MTP should not be in the first pp rank."
```
Comment (asolergi-nv) [inline megatron/core/transformer/pipeline_parallel_layer_layout.py]: _Currently_

File: tests/unit_tests/pipeline_parallel/test_pipeline_layout.py [hunk 3]
```diff
@@ -81,6 +84,10 @@ def initialize_gpt_model(
             )
         else:
             mtp_block_spec = None
+
+        # print("========================")
+        # print("[DEBUG] mtp_block_spec is ", mtp_block_spec)
+        # exit()
```
Comment (asolergi-nv) [inline tests/unit_tests/pipeline_parallel/test_pipeline_layout.py]:
```suggestion
```

File: megatron/core/transformer/multi_token_prediction.py [hunk 4]
```diff
@@ -332,25 +332,103 @@ def get_mtp_layer_spec_for_backend(
...
+    return list(mtp_ranks)
+
+
+def get_mtp_layer_offset(config: TransformerConfig, vp_stage: Optional[int] = None) -> int:
     """Get the offset of the MTP layer."""
-    # Currently, we only support put all of MTP layers on the last pipeline stage.
-    return 0
+    # TODO(shifangx): Currently, we only support put all of MTP layers
+    # on the last pipeline stage, so the offset is always 0.
+    # We will support more flexible MTP placement in the future.
```
Comment (asolergi-nv) [inline megatron/core/transformer/multi_token_prediction.py]: ¿?

File: megatron/training/training.py [hunk 5]
```diff
@@ -2821,18 +2821,20 @@ def get_train_valid_test_num_samples():
     return (train_samples, eval_samples, test_iters * args.global_batch_size)
 
 
-def build_train_valid_test_datasets(build_train_valid_test_datasets_provider, train_valid_test_num_samples=None):
+def build_train_valid_test_datasets(build_train_valid_test_datasets_provider, train_valid_test_num_samples=None, vp_stage=None):
     """Build pretraining datasets."""
     if train_valid_test_num_samples is None:
         train_valid_test_num_samples = get_train_valid_test_num_samples()
-    print_rank_0(' > datasets target sizes (minimum size):')
```
Comment (asolergi-nv) [inline megatron/training/training.py:2946]: Why would we drop this message?

File: megatron/training/training.py [hunk 6]
```diff
@@ -2821,18 +2821,20 @@ def get_train_valid_test_num_samples():
...
     print_rank_0('    train:      {}'.format(train_valid_test_num_samples[0]))
     print_rank_0('    validation: {}'.format(train_valid_test_num_samples[1]))
     print_rank_0('    test:       {}'.format(train_valid_test_num_samples[2]))
-    return build_train_valid_test_datasets_provider(train_valid_test_num_samples)
+    if vp_stage is not None:
+        return build_train_valid_test_datasets_provider(train_valid_test_num_samples, vp_stage=vp_stage)
+    else:
+        return build_train_valid_test_datasets_provider(train_valid_test_num_samples)
```
Comment (asolergi-nv) [inline megatron/training/training.py]: Do we really need this?

File: megatron/core/transformer/pipeline_parallel_layer_layout.py (hunk 2 above)
Comment (BestJuly) [reply megatron/core/transformer/pipeline_parallel_layer_layout.py]: Good catch! Thanks for your review and I have modified the code according to your review comments. The UT should pass and I have checked the failed one which also passed with the CI container. I believe that would be a environment issue.
```
============================================================================= short test summary info =============================================================================
PASSED tests/unit_tests/data/test_builder.py::test_builder
```
May I have your further review then? Thank you.

**Comment by asolergi-nv** [comment]
/ok to test [d03c64c](https://github.com/ko3n1g/Megatron-LM/commit/d03c64c8a9a785a79c14170f18076dbf13d11e87)

**Review by asolergi-nv — APPROVED** [review]
LGTM

**Review by jkamalu — APPROVED** [review]
